### PR TITLE
Add slow shader post

### PR DESCRIPTION
**Pull request by raphlinus**
This is the first draft, needs performance numbers, a tiger graphic, and
probably more polish overall.

### CODE REVIEW

File: _posts/2021-04-22-slow-shader.md [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+
+Overall, I've found the Adreno 640 to be generally similar to low-end desktop GPUs, at least in terms of the number of registers available per thread, scheduling of subgroups within a processing cluster, etc. There are fewer processing clusters, a slower clock speed, and less memory bandwidth overall, but those should scale fairly smoothly with existing code and are entirely to be expected. There's more detail about the [A6xx Streaming Processor][A6xx SP] at the Freedreno wiki, and my empirical observations are consistent.
+
+## Conclusion and thanks
+
+Developing GPU compute shaders risks mysterious performance slowdowns when deploying across a range of hardware. With some determination and tools, it's possible to diagnose those. Fortunately, there's a lot of activity on better open source tools, much of which is being done at Google, but of course the broader open source graphics community deserves credit too.
```
Comment (dneto0) [inline _posts/2021-04-22-slow-shader.md:111]: Suggest

> Fortunately, open source tools are always getting better providing insight into what's happening inside the machine.  I'm proud that fellow Googler's are very active in this space, in collaboration with the broader open source graphics community.

File: _posts/2021-04-22-slow-shader.md [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+
+One such slowdown is overly conservative memory coherency. In the future, the Vulkan memory model promises to make these performance concerns more explicit and predictable - a slowdown when you're explicitly asking for barriers or coherent semantics would not be surprising. In the meantime, it's something to be aware of and test. In particular, just because performance is good on one set of GPUs and drivers is not a guarantee it won't be a problem on others.
+
+Thanks to Rob Clark for much valuable insight about Adreno ISA, Lei Zhang for pointing me to many useful Adreno (and mobile GPU resources), Hugues Hevrard for help with AGI, and Elias Naur for going on this journey with me, as well as tool support in the form of making glGetProgramBinary easy to run.
```
Comment (dneto0) [inline _posts/2021-04-22-slow-shader.md:115]: "Hugues Evrard"

**Review by dneto0 — COMMENTED** [review]
Nice writeup!

File: _posts/2021-04-22-slow-shader.md [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+
+Fortunately, for someone as determined as I was, it's possible, largely thanks to the existence of free tools, especially [Freedreno]. I was prepared to go as deep as I needed, including getting Freedreno working on some hardware; if it performed well, that would point to the proprietary driver. But if the performance matched, I would easily be able to see the ISA it produced, and dig into that.
+
+As it turned out, I didn't have to go that far. The [glGetProgramBinary] function can be used to spit out a binary, and then the [disassembly tools] in Mesa could spit out disassembler for me to read. There's a bit more complexity to this, as the drivers produce both a compressed and an uncompressed format. The corresponding Vulkan call is [vkGetPipelineCacheData], but I was only able to get the compressed form. I strongly suspect it's possible to work out the compression format, but when I got readable results from the GL route, that was good enough for me.
```
[the remark this answers is not in the document]
Comment (critsec) [reply _posts/2021-04-22-slow-shader.md:69]: My reading of this is that the binary was extracted from the proprietary driver, and then Mesa tools were used offline to disassemble it. That avoided the question of whether differences between the proprietary and freedreno compilers and drivers were significant (at least in this particular case).

File: _posts/2021-04-22-slow-shader.md [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+
+Given this knowledge, a fix is fairly straightforward, and we also know how to avoid this particular problem in the future. Most likely, we'll just use scratch memory for the clip stack, rather than sharing a buffer with the scene and path, so all access to that buffer is readonly. There are plenty of other solutions that would work, including a dedicated buffer for the clip stack.
+
+## Why we need the Vulkan memory model
+
+In the middle of the investigation, I viewed the proprietary shader compiler with extreme suspicion: it's making my code run slow, for no good reason that I could see, and also not matching the performance expectations set by other (desktop) GPUs. But now I understand much better why it's doing that.
+
+Even so, in the longer term I think it's possible to do much better, and I believe the key to that is the [Vulkan memory model]. (I've written about this before in my [prefix sum][Prefix sum on Vulkan] blog post, but at that time I did not observe a substantial performance difference and this time I do)
```
Comment (critsec) [inline _posts/2021-04-22-slow-shader.md:95]: typo: missing period for last sentence.

File: _posts/2021-04-22-slow-shader.md [hunk 5]
```diff
@@ -0,0 +1,132 @@
+---
+layout: post
+title:  "The case of the curiously slow shader"
+date:   2021-04-22 07:53:42 -0700
+categories: [gpu]
+---
+These, days a significant amount of my time and energy is getting [piet-gpu], a compute-focused 2D rendering engine, to run well on mobile hardware. Not too long ago, I got it running on Pixel 4, and breathlessly waited for the performance numbers, which turned out to be... disappointing. I was able to figure out why, but therein lies a story.
```
Comment (ShabbyX) [inline _posts/2021-04-22-slow-shader.md:7]: s/These, days/These days/

File: _posts/2021-04-22-slow-shader.md [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+
+As it turned out, I didn't have to go that far. The [glGetProgramBinary] function can be used to spit out a binary, and then the [disassembly tools] in Mesa could spit out disassembler for me to read. There's a bit more complexity to this, as the drivers produce both a compressed and an uncompressed format. The corresponding Vulkan call is [vkGetPipelineCacheData], but I was only able to get the compressed form. I strongly suspect it's possible to work out the compression format, but when I got readable results from the GL route, that was good enough for me.
+
+Looking at the disassembly (the full traces are in the [piet-gpu bug]), one thing immediately stuck out. In the happy path, all reads from memory use the `isam` (Image SAMple) instruction, like this:
+
+```
+:5:0166:0205[a0005100x_02000001x] isam (s32)(x)r0.x, r0.x, s#0, t#1
```
Comment (ShabbyX) [inline _posts/2021-04-22-slow-shader.md:74]: Suggest dropping stuff before `isam` to reduce unnecessary clutter. Seam with the `ldib` instruction below.

File: _posts/2021-04-22-slow-shader.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (raphlinus) [reply _posts/2021-04-22-slow-shader.md:69]: I added a little language that I hope clarifies this even further, that I used free tools to gain more introspection into the results with the proprietary driver.

File: _posts/2021-04-22-slow-shader.md [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+
+In this particular case, the shader needs only the most relaxed possible memory semantics; each memory write is consumed only by a read on the same thread, which means no special semantics are necessary. Without taking the Vulkan memory model into account, the compiler has no way to know that. The proprietary compiler *does* [report][Pixel 4 at vulkan.gpuinfo.org] that it respects the Vulkan memory model, but it's not lying; if it always generates conservative code and then ticks the "support" bit, it's satisfying the requirements. What is fair to say is that it's missing an opportunity to optimize. Even so, I understand why they haven't prioritized it. When I researched my prefix sum blog post, I didn't find any evidence people were actually writing shaders that used the Vulkan memory model, and I haven't heard of any since.
```
[the remark this answers is not in the document]
Comment (raphlinus) [reply _posts/2021-04-22-slow-shader.md:103]: I see your point, but am not convinced it's extremely relevant. For a thread not to coherently see its own writes would, I think, tank performance almost to the point of unusability. Just to take one example, you wouldn't be able to spill registers without putting barriers or whatnot into place.

I think there's a general point, which is that relaxed read/writes are not *necessarily* going to be as fast as a readonly memory, and I think I'll add language to that effect. I don't think it's worth going into a lot more detail than that.

**Pull request merged**